File: mood_diary.js
// mood_diary.js
const fs = require('fs').promises;
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (prompt) => new Promise(resolve => rl.question(prompt, resolve));

class MoodEntry {
    constructor(id, mood, note, timestamp = new Date().toISOString()) {
        this.id = id;
        this.timestamp = timestamp;
        this.mood = mood;
        this.note = note;
    }
}

class MoodDiary {
    constructor() {
        this.entries = [];
        this.nextId = 1;
    }

    addEntry(mood, note) {
        if (mood < 1 || mood > 10) {
            throw new Error('Оценка должна быть от 1 до 10');
        }
        const entry = new MoodEntry(this.nextId, mood, note);
        this.entries.push(entry);
        this.nextId++;
        return entry;
    }

    findEntry(id) {
        return this.entries.find(e => e.id === id);
    }

    deleteEntry(id) {
        const index = this.entries.findIndex(e => e.id === id);
        if (index === -1) return false;
        this.entries.splice(index, 1);
        return true;
    }

    getStats() {
        const total = this.entries.length;
        if (total === 0) {
            return { total: 0, average: null, min: null, max: null };
        }
        const moods = this.entries.map(e => e.mood);
        const sum = moods.reduce((a, b) => a + b, 0);
        return {
            total,
            average: sum / total,
            min: Math.min(...moods),
            max: Math.max(...moods)
        };
    }

    async saveToFile(filename = 'mood_data.json') {
        await fs.writeFile(filename, JSON.stringify(this.entries, null, 2));
    }

    async loadFromFile(filename = 'mood_data.json') {
        try {
            const data = await fs.readFile(filename, 'utf8');
            const loaded = JSON.parse(data);
            this.entries = loaded.map(e => Object.assign(new MoodEntry(0), e));
            this.nextId = this.entries.reduce((max, e) => Math.max(max, e.id), 0) + 1;
        } catch (err) {
            if (err.code !== 'ENOENT') throw err;
        }
    }
}

function printEntry(entry) {
    const emoji = entry.mood >= 7 ? '😄' : entry.mood >= 4 ? '😐' : '😞';
    console.log(`${emoji} #${entry.id} - ${entry.timestamp} - Оценка: ${entry.mood}/10`);
    console.log(`   Заметка: ${entry.note}`);
}

async function main() {
    const diary = new MoodDiary();
    await diary.loadFromFile();

    while (true) {
        console.log('\n===== ДНЕВНИК НАСТРОЕНИЯ (JavaScript) =====');
        console.log('1. Добавить запись');
        console.log('2. Показать все записи');
        console.log('3. Показать статистику');
        console.log('4. Удалить запись');
        console.log('5. Сохранить в файл');
        console.log('6. Загрузить из файла');
        console.log('0. Выход');
        const choice = await question('Выберите действие: ');

        if (choice === '0') break;

        switch (choice) {
            case '1': {
                const mood = parseInt(await question('Оценка настроения (от 1 до 10): '));
                if (isNaN(mood) || mood < 1 || mood > 10) {
                    console.log('Оценка должна быть от 1 до 10.');
                    continue;
                }
                const note = await question('Заметка: ');
                try {
                    const entry = diary.addEntry(mood, note);
                    console.log(`Запись добавлена с ID ${entry.id}`);
                } catch (err) {
                    console.log(err.message);
                }
                break;
            }
            case '2': {
                if (diary.entries.length === 0) {
                    console.log('Нет записей.');
                } else {
                    diary.entries.forEach(printEntry);
                }
                break;
            }
            case '3': {
                const stats = diary.getStats();
                if (stats.total === 0) {
                    console.log('Нет записей для статистики.');
                } else {
                    console.log('\n=== СТАТИСТИКА ===');
                    console.log(`Всего записей: ${stats.total}`);
                    console.log(`Средняя оценка: ${stats.average.toFixed(2)}`);
                    console.log(`Минимальная оценка: ${stats.min}`);
                    console.log(`Максимальная оценка: ${stats.max}`);
                }
                break;
            }
            case '4': {
                const id = parseInt(await question('Введите ID записи для удаления: '));
                if (diary.deleteEntry(id)) {
                    console.log('Запись удалена.');
                } else {
                    console.log('Запись не найдена.');
                }
                break;
            }
            case '5':
                try {
                    await diary.saveToFile();
                    console.log('Сохранено.');
                } catch (err) {
                    console.log('Ошибка сохранения:', err.message);
                }
                break;
            case '6':
                try {
                    await diary.loadFromFile();
                    console.log('Загружено.');
                } catch (err) {
                    console.log('Ошибка загрузки:', err.message);
                }
                break;
            default:
                console.log('Неизвестная команда.');
        }
    }
    rl.close();
}

main().catch(console.error);

File: mood_diary.go
// mood_diary.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type MoodEntry struct {
	ID        int       `json:"id"`
	Timestamp time.Time `json:"timestamp"`
	Mood      int       `json:"mood"`
	Note      string    `json:"note"`
}

type MoodDiary struct {
	entries []MoodEntry
	nextID  int
}

func NewMoodDiary() *MoodDiary {
	return &MoodDiary{
		entries: []MoodEntry{},
		nextID:  1,
	}
}

func (d *MoodDiary) AddEntry(mood int, note string) (MoodEntry, error) {
	if mood < 1 || mood > 10 {
		return MoodEntry{}, fmt.Errorf("оценка должна быть от 1 до 10")
	}
	entry := MoodEntry{
		ID:        d.nextID,
		Timestamp: time.Now(),
		Mood:      mood,
		Note:      note,
	}
	d.entries = append(d.entries, entry)
	d.nextID++
	return entry, nil
}

func (d *MoodDiary) FindEntry(id int) *MoodEntry {
	for i := range d.entries {
		if d.entries[i].ID == id {
			return &d.entries[i]
		}
	}
	return nil
}

func (d *MoodDiary) DeleteEntry(id int) bool {
	for i, e := range d.entries {
		if e.ID == id {
			d.entries = append(d.entries[:i], d.entries[i+1:]...)
			return true
		}
	}
	return false
}

func (d *MoodDiary) GetStats() map[string]interface{} {
	total := len(d.entries)
	if total == 0 {
		return map[string]interface{}{"total": 0, "average": nil, "min": nil, "max": nil}
	}
	sum := 0
	min := 11
	max := 0
	for _, e := range d.entries {
		sum += e.Mood
		if e.Mood < min {
			min = e.Mood
		}
		if e.Mood > max {
			max = e.Mood
		}
	}
	return map[string]interface{}{
		"total":   total,
		"average": float64(sum) / float64(total),
		"min":     min,
		"max":     max,
	}
}

func (d *MoodDiary) SaveToFile(filename string) error {
	data, err := json.MarshalIndent(d.entries, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, data, 0644)
}

func (d *MoodDiary) LoadFromFile(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var loaded []MoodEntry
	if err := json.Unmarshal(data, &loaded); err != nil {
		return err
	}
	d.entries = loaded
	for _, e := range loaded {
		if e.ID >= d.nextID {
			d.nextID = e.ID + 1
		}
	}
	return nil
}

func readString(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func readInt(prompt string) int {
	for {
		input := readString(prompt)
		if val, err := strconv.Atoi(input); err == nil {
			return val
		}
		fmt.Println("Введите число.")
	}
}

func printEntry(entry MoodEntry) {
	emoji := "😄"
	if entry.Mood >= 7 {
		emoji = "😄"
	} else if entry.Mood >= 4 {
		emoji = "😐"
	} else {
		emoji = "😞"
	}
	fmt.Printf("%s #%d - %s - Оценка: %d/10\n", emoji, entry.ID, entry.Timestamp.Format("2006-01-02 15:04:05"), entry.Mood)
	fmt.Printf("   Заметка: %s\n", entry.Note)
}

func main() {
	diary := NewMoodDiary()
	if err := diary.LoadFromFile("mood_data.json"); err != nil {
		fmt.Println("Ошибка загрузки:", err)
	}

	for {
		fmt.Println("\n===== ДНЕВНИК НАСТРОЕНИЯ (Go) =====")
		fmt.Println("1. Добавить запись")
		fmt.Println("2. Показать все записи")
		fmt.Println("3. Показать статистику")
		fmt.Println("4. Удалить запись")
		fmt.Println("5. Сохранить в файл")
		fmt.Println("6. Загрузить из файла")
		fmt.Println("0. Выход")
		choice := readString("Выберите действие: ")

		switch choice {
		case "0":
			return
		case "1":
			mood := readInt("Оценка настроения (от 1 до 10): ")
			if mood < 1 || mood > 10 {
				fmt.Println("Оценка должна быть от 1 до 10.")
				continue
			}
			note := readString("Заметка: ")
			entry, err := diary.AddEntry(mood, note)
			if err != nil {
				fmt.Println(err)
				continue
			}
			fmt.Printf("Запись добавлена с ID %d\n", entry.ID)
		case "2":
			if len(diary.entries) == 0 {
				fmt.Println("Нет записей.")
			} else {
				for _, e := range diary.entries {
					printEntry(e)
				}
			}
		case "3":
			stats := diary.GetStats()
			if stats["total"] == 0 {
				fmt.Println("Нет записей для статистики.")
			} else {
				fmt.Println("\n=== СТАТИСТИКА ===")
				fmt.Printf("Всего записей: %d\n", stats["total"])
				fmt.Printf("Средняя оценка: %.2f\n", stats["average"])
				fmt.Printf("Минимальная оценка: %d\n", stats["min"])
				fmt.Printf("Максимальная оценка: %d\n", stats["max"])
			}
		case "4":
			id := readInt("Введите ID записи для удаления: ")
			if diary.DeleteEntry(id) {
				fmt.Println("Запись удалена.")
			} else {
				fmt.Println("Запись не найдена.")
			}
		case "5":
			if err := diary.SaveToFile("mood_data.json"); err != nil {
				fmt.Println("Ошибка сохранения:", err)
			} else {
				fmt.Println("Сохранено.")
			}
		case "6":
			if err := diary.LoadFromFile("mood_data.json"); err != nil {
				fmt.Println("Ошибка загрузки:", err)
			} else {
				fmt.Println("Загружено.")
			}
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}
